projects/tomlkit/atheris_tomlkit_dict.py

```diff
@@ -0,0 +1,149 @@
+# Copyright 2023 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+################################################################################
+
+# This is a modified version of https://github.com/robtandy/randomdict which has
+# been changed to
+#    - Work with the atheris fuzzer
+#    - Filter duplicate keys
+
+import datetime
+from itertools import product
+from random import choice, shuffle
+from typing import Callable, List
+
+import tomlkit
+
+fdp = None
+
+
+def random_string():
+    return fdp.ConsumeString(fdp.ConsumeIntInRange(0, 1024))
+
+
+def random_int():
+    return fdp.ConsumeInt(16)
+
+
+def random_bool():
+    return fdp.ConsumeBool()
```

Comment:
I find this to be a bit overhead-ish -- what's the reason for not just using `fdp` directly?

Reply:
It was originally to stay in keeping with the style of the original random dictionary generator this was derived from. I think it's better to keep these `random_bool()`, `random_int()` etc methods as we use them later on for logic around the random dictionary generation. For example in this logic to exclude certain generator methods from being used for dictionary keys.

``` python
if key_gen not in [
      random_bool,
      ...
```

Reply:
> as we use them later on for logic around the random dictionary generation.

Is it possible to do `fdp.ConsumeBool` as a key, or can functions only be keys when they are not affiliated with an object? If possible then I'd probably lean towards keeping `fdp` functions rather than the wrappers

Reply:
I think this is tricky, I've had a go and I think you move the complexity around by getting rid of the `generate_*` boilerplate. And as you say there's the additional trickiness of can functions only be compared and used as generators if they're not affiliated with an object (I'm not sure currently).

Also for the change above I need to call the `ConsumeBool` method on the `fdp` object as part of generating the random dictionary. Currently when generating the dictionary I assume I can call the different `generate_*` methods directly without and additional arguments or logic, e.g. `key_gen()` and `val_gen()`. If I was to start mixing pure atheris calls (e.g. `ConsumeBool`) and calls where I need to do some additional logic (e.g. `random_datetime` because atheris doesn't provide a native random datetime) then I'd need to write additional code to call the generator method correctly (e.g. `ConsumeBool` should be called against an `fdp` object, `generate_datetime` can be called directly).

For interest, in the background I'm working on re-building this random dictionary generator implementation into a library that can accept atheris as it's source https://github.com/sg3-141-592/dictgen (still work in progress) as it seems like there's a few projects this would be useful for (json, toml, yaml dumpers etc). And I can get rid of these ad-hoc `atheris_dict.py`, `atheris_tomlkit_doc.py`s that are around the place. This random dictionary method was effective on toml ([toml - introspector report - 15-04-2023](https://storage.googleapis.com/oss-fuzz-introspector/toml/inspector-report/20230415/fuzz_report.html)) so seems like an area worth investing in.